### PR TITLE
WIP: Current user in rules

### PR DESCRIPTION
**Pull request by fedemoretti**
First approach to using current_user inside rules blocks.

Please let me know what you think about this. It is based on the work started by @tute in #100

### CODE REVIEW

**Comment by fedemoretti** [comment]
Ok, branch seems to be a little old. I'll fix the conflicts and update it.

**Comment by fedemoretti** [comment]
Updated with master. Does this PR make any sense?

**Comment by tute** [comment]
This is looking good! I like very much the fact that it resulted backwards compatible. Will give detailed feedback later in the week. Thank you!

File: README.md [hunk 1]
```diff
@@ -112,6 +113,17 @@ end
 grant_on ['users#create', 'users#update'], badge: 'autobiographer', temporary: true do |user|
   user.name? && user.email?
 end
+
+grant_on 'users#update', badge: 'updated_my_info' do |user, current_user:|
+  user == current_user
+end
```
Comment (tute) [inline README.md]: This example can get confusing as we speak of `user` in two different contexts. Let's do:

```ruby
grant_on 'addresses#update', badge: 'address_updated' do |address, current_user:|
  current_user == address.user
end
```

File: README.md [hunk 2]
```diff
@@ -112,6 +113,17 @@ end
...
+```
+
+## Even only using the current user
```
Comment (tute) [inline README.md]: Suggestion: "You may use current_user only".

Also, add a new line after the title please?

File: lib/merit/rule.rb [hunk 3]
```diff
@@ -7,19 +7,14 @@ class Rule
                   :multiple, :temporary, :score, :block, :category
 
     # Does this rule's condition block apply?
-    def applies?(target_obj = nil)
+    def applies?(target_obj = nil, current_user = nil)
```
Comment (tute) [inline lib/merit/rule.rb]: Both of these ` = nil` seem like a bug to me. I think these arguments should always be required.

File: lib/merit/rule.rb [hunk 4]
```diff
@@ -7,19 +7,14 @@ class Rule
...
-          block.call(target_obj)
-        else
-          Rails.logger.warn '[merit] no target_obj found on Rule#applies?'
-          false
-        end
-      when 0
-        block.call
+      if target_obj.nil? && current_user.nil?
+        Rails.logger.warn '[merit] no target_obj
+                          or current_user found on Rule#applies?'
```
Comment (tute) [inline lib/merit/rule.rb]: I'd check for each one separately:

```ruby
if target_obj.nil?
  Rails.logger.warn '[merit] no target_object found on Rule#applies?'
  return false
end
if current_user.nil?
  Rails.logger.warn '[merit] no current_user found on Rule#applies?'
  return false
end

call_rule_block(target_obj, current_user)
```
While we are on this, can we rename `target_obj` to `target_object` in this file please?

Comment (tute) [inline lib/merit/rule.rb]: Actually, these checks depend on the arity of the rule block. I think we'll need to rework a bit this class.

**Comment by tute** [comment]
Looking good, thank you for your work! :clap:

**Comment by tute** [comment]
Thanks @fedemoretti! Will you apply the PR feedback? Do you need any help doing so?

**Comment by fedemoretti** [comment]
@tute I just fixed some of the comments (the simple ones), 'cause I still have some doubts about the ones in `Rule` class.
I've been quite busy these days, but I'll try to get back to you during the weekend with my thoughts.
Thanks for your quick response and comments :)

**Comment by tute** [comment]
> I still have some doubts about the ones in Rule class.

Yes, me too. If you'd like to pair on this, I can probably find sometime in two weeks.

**Comment by tute** [comment]
We can pair on this in December if you'd like, @fedemoretti. I'll have time for wrapping this up.

**Comment by fedemoretti** [comment]
Sure! We'll finish this up.

**Comment by tute** [comment]
Closing due to lack of activity.

**Comment by Ldaruu** [comment]
At the end this feature didn't get merged, right? @tute

**Pull request closed without merging**